astroquery/gemini/core.py

```diff
@@ -0,0 +1,480 @@
+            url = "%s/%s" % (url, adaptive_optics)
+        if program_text is not None:
+            url = "%s/ProgramText=%s" % (url, program_text)
+        if objectname is not None:
+            url = "%s/object=%s" % (url, objectname)
+        if raw_reduced is not None:
+            if raw_reduced not in __valid_raw_reduced__:
+                raise ValueError("Unrecognized raw/reduced setting: %s" % raw_reduced)
+            url = "%s/%s" % (url, raw_reduced)
+        response = self._request(method="GET", url=url, data={}, timeout=180, cache=False)
+
+        js = json.loads(response.text)
+        return _gemini_json_to_table(js)
+
+    @class_or_instance
+    def query_raw(self, *args, **kwargs):
+        """
+        perform flexible query against Gemini observations
+
+        This is a more flexible query method.  This method will do special handling for
+        coordinates and radius if present in kwargs.  However, for the remaining arguments
+        it assumes all of args are useable as query path elements.  For kwargs, it assumes
+        all of the elements can be passed as name=value within the query path to Gemini.
+
+        This method does not do any validation checking or attempt to interperet the
+        values being passed, aside from coordinates and radius.
+
+        This method is most useful when the query_criteria and query_region do not
+        meet your needs and you can build the appropriate search in the website.  When
+        you see the URL that is generated by the archive, you can translate that into
+        an equivalent python call with this method.  For example, if the URL in the
+        website is:
+
+        https://archive.gemini.edu/searchform/RAW/cols=CTOWEQ/notengineering/GMOS-N/PIname=Hirst/NotFail
+
+        You can disregard NotFail, cols=x and notengineering.  You would run this query as
+
+        query_raw('GMOS-N', PIname='Hirst')
+
+        Parameters
+        ----------
+        args :
+            The list of parameters to be passed via the query path to the webserver
+        kwargs :
+            The dictionary of parameters to be passed by name=value within the query
+            path to the webserver
+
+        Returns
+        -------
+        response : `~astropy.table.Table`
+        """
+        if 'radius' in kwargs:
+            radius = kwargs['radius']
+        else:
+            radius = None
+        if radius is not None:
+            if isinstance(radius, (int, float)):
+                radius = radius * units.deg
+            radius = astropy.coordinates.Angle(radius)
+        if 'coordinates' in kwargs:
+            coordinates = kwargs['coordinates']
+            coordinates = commons.parse_coordinates(coordinates)
+        else:
+            coordinates = None
+
+        if coordinates is not None:
+            url = "%s/ra=%f/dec=%f" % (url, coordinates.ra.deg, coordinates.dec.deg)
+        if radius is not None:
+            url = "%s/sr=%fd" % (url, radius.deg)
```

Comment:
there must be a supported way to construct these sorts of URLs in python, but I can't find it.  Just an aside - if you can find a library to do this for you, it would be much nicer than these sequential `if` statements.

Reply:
hmm - if I made the arguments match the REST service's expectations then this would drop out and be handled by the code at the end as generic kw/value pairs.  Though the user would also have to pass the radius with the trailing 'd' to get degrees.  Otherwise the logic to convert from coordinates/radius to this special sauce has to be captured somewhere.  It could also be placed in a dictionary of handlers outside the method and then the loop could just be "if kw in handler_dict, use handler_dict[kw] to modify url, else use default .../kw=val/"

Reply:
see comment on following thread, tried an approach separating out the url logic